utils/networking/src/network_address.rs

```diff
@@ -0,0 +1,302 @@
+// Copyright (c) 2021-2024 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    fmt::Display,
+    net::{IpAddr, SocketAddr},
+    str::FromStr,
+};
+
+use itertools::Either;
+use serde_with::{DeserializeFromStr, SerializeDisplay};
+use thiserror::Error;
+
+use crate::{
+    ip_or_socket_address::IpOrSocketAddress,
+    resolvable_name::{NameResolutionError, ResolvableName},
+};
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct DomainName(String);
```

Comment:
I've been back and forth in my mind to whether I should say this or just let it go... but the correct name is probably Hostname, not DomainName... the reason is that a domain is usually a superset of what hostnames are or where they are. Maybe I'm missing something, I don't know. But I find the name Hostname much more appropriate for a simple name that's resolvable using a DNS server.

Reply:
https://www.baeldung.com/wp-content/uploads/sites/4/2022/02/szfafsd.drawio.png

Reply:
[discussed] I've left it as is.